### ISSUE TITLE
Spack setup-env.sh is really heavy

### ISSUE THREAD

**Issue by ashwinvis** [spack/spack#7355]
### Summary

Sourcing the line `. $SPACK_ROOT/share/spack/setup-env.sh` in the `.zshrc` adds to alteast 1 second of startup time. It can become annoying, everytime you open a new tab or activate a new virtualenv for python.

#### What I expected

< 0.2 second of added startup time is reasonable.

#### What happened instead

```console
# With the setup-env.sh line
~ » time zsh -ic "source ~/.zshrc"
zsh -ic "source ~/.zshrc"  1.70s user 0.46s system 99% cpu 2.159 total

# Without the setup-env.sh line
~ » time zsh -ic "source ~/.zshrc"
zsh -ic "source ~/.zshrc"  0.43s user 0.18s system 99% cpu 0.611 total
```

See where the time is spent inside setup-env.sh
```console
# Commenting out the "build and make available environment-modules" block 
~ » time zsh -ic "source ~/.zshrc"
zsh -ic "source ~/.zshrc"  1.31s user 0.40s system 92% cpu 1.850 total

# Commenting out the "Set up modules and dotkit search paths in the user environment" block
~ » time zsh -ic "source ~/.zshrc"
zsh -ic "source ~/.zshrc"  0.50s user 0.31s system 100% cpu 0.797 total
```
### Information on your system

 1. Linux 64-bit (Arch Linux). Solid state hard disk
 2. `spack bootstrap` done.

**Comment by ashwinvis** [comment]
1st block 

https://github.com/spack/spack/blob/f302d642358e28ed7a4efc1a3ccd9dc0bba604fb/share/spack/setup-env.sh#L221-L234

2nd block

https://github.com/spack/spack/blob/f302d642358e28ed7a4efc1a3ccd9dc0bba604fb/share/spack/setup-env.sh#L236-L250

**Comment by alalazo** [comment]
The problems here are the calls to the `python` interpreter, which requires a fixed start-up time. A couple of months ago I joined the calls in the second block (there were three as far as I remember). We can probably skip both blocks on user request - e.g. if one doesn't need modules.

**Comment by ashwinvis** [comment]
Well if one doesn't need modules, adding `$SPACK_ROOT/bin` to `PATH` would suffice. I don't know how these python calls are implemented but can it be made faster by:
 
 1. substituting them with a shell equivalent
 1. using threads

**Comment by sethrj** [comment]
If you don't need module or dotkit support, and if you're using bash, I think you can replace the call to `setup-env` with:
```
export SPACK_ROOT=/path/to/spack
export PATH=${SPACK_ROOT}/bin:${PATH}
source ${SPACK_ROOT}/share/spack/spack-completion.bash
```

**Comment by ashwinvis** [comment]
@sethrj I understand, but then you cannot do `spack load` or `module load` and you use spack the way it is meant to be. The only speedup in loading is by a hack which looks like:

```sh
export SPACK_ROOT=/path/to/spack
export PATH=$SPACK_ROOT/bin/:$PATH

printf 'Added spack to PATH and loading environment-modules\n'
export MODULEPATH=$SPACK_ROOT/share/spack/modules/linux-archrolling-x86_64
export MODULE_PREFIX=$SPACK_ROOT/opt/spack/linux-archrolling-x86_64/gcc-7.2.1/environment-modules-3.2.10-7ut5ux3rkkw5z3v6rnr6mfmfjas7px6b
module () {
	eval `${MODULE_PREFIX}/Modules/bin/modulecmd zsh $*`
}
```

which allows me to use `module load ` atleast.

**Comment by alalazo** [comment]
The start-up time improved a lot over the last couple of years. Might be further improved after #14062

**Comment by sethrj** [comment]
Even though the startup time is much improved, it's an extra delay when launching *every* new interactive terminal session, even though the likelihood of using `spack` in any particular session is small.

My solution is to lazily load the spack environment by putting this in my `.bashrc`:

```sh
spack() {
  unset spack
  source ${SPACK_ROOT}/share/spack/setup-env.sh
  spack "$@"
}
```

The only downside to this is that spack autocompletion won't before the first invocation of `spack`. In practice I haven't found this to be an issue.

**Comment by alalazo** [comment]
> Even though the startup time is much improved, it's an extra delay when launching every new interactive terminal session, even though the likelihood of using spack in any particular session is small.

#14062 might give us the possibility to remove the call to `spack` that is done during the start-up. At that point we won't have to wait for the Python interpreter to start (which is what is causing most of the delay) and we'll be evaluating only shell code. Hope that will be sufficient to close this issue, but we'll see.

**Comment by adamjstewart** [comment]
This should be fixed by #24545 and #24236. Users who don't use Spack's module features can set `SPACK_SKIP_MODULES` and skip all the slow calls to the Python interpreter.
